### PR TITLE
Adding Bulk feature to Qed

### LINKED ISSUE #105: Feature: bulk inserts
Tasks:
- [x] Api: new handler.
- [x] Api: new protocol definitions.
- [x] Client: AddBulk
- [x] Wal: new command.
- [x] Wal: Apply method.
- [x] Fsm: Apply method.
- [x] Fsm: Benchmarking.
- [x] Balloon: Apply method.
- [x] Balloon: New struct "snapshotBulk".
- [x] Balloon: Benchmarking.
- [x] Riot: update.
- [ ] Improve error handling.

### PR DESCRIPTION
**Pull request by iknite**
Closes #105

### CODE REVIEW

File: balloon/balloon.go [hunk 1]
```diff
@@ -226,11 +232,70 @@ func (b *Balloon) Add(event []byte) (*Snapshot, []*storage.Mutation, error) {
...
+		snapshotBulk.Snapshots = append(snapshotBulk.Snapshots, Snapshot{
+			EventDigest:   eventBulkDigest[i],
+			HistoryDigest: historyDigests[i],
+			HyperDigest:   hyperDigest,
+			Version:       eventVersions[i],
+		})
+	}
+
+	// Increment version
+	stats.Set("version", metrics.Uint64ToVar(b.version))
```
Comment (aalda) [inline balloon/balloon.go]: See my comment above

File: balloon/balloon.go [hunk 2]
```diff
@@ -74,6 +74,12 @@ type Snapshot struct {
 	Version       uint64
 }
 
+// BulkSnapshot is the struct that has both history and hyper digest and the
+// current version for that rootNode digests.
+type SnapshotBulk struct {
```
Comment (aalda) [inline balloon/balloon.go]: Why didn't use a type like `type SnapshotBulk []Snapshot`?

File: balloon/balloon_test.go [hunk 3]
```diff
@@ -54,7 +54,40 @@ func TestAdd(t *testing.T) {
...
+		[]byte("Morning's at seven;"),
+		[]byte("The hill-side’s dew-pearled;"),
+		[]byte("The lark's on the wing;"),
+		[]byte("The snail's on the thorn;"),
+		[]byte("God's in his heaven—"),
+		[]byte("All's right with the world!"),
+	}
+
+	snapshotBulk, mutations, err := balloon.AddBulk(events)
+	store.Mutate(mutations)
```
Comment (aalda) [inline balloon/balloon_test.go]: Always check Mutate errors

File: protocol/events.go [hunk 4]
```diff
@@ -33,6 +33,10 @@ type Event struct {
 	Event []byte
 }
 
+type EventBulk struct {
```
Comment (aalda) [inline protocol/events.go]: Rename to EventsBulk?

File: protocol/events.go [hunk 5]
```diff
@@ -49,10 +53,15 @@ type MembershipDigest struct {
...
 type Snapshot struct {
+	EventDigest   hashing.Digest
 	HistoryDigest hashing.Digest
 	HyperDigest   hashing.Digest
 	Version       uint64
-	EventDigest   hashing.Digest
+}
+
+// SnapshotBulk is the public struct that apihttp.AddBulk Handler call returns.
+type SnapshotBulk struct {
```
Comment (aalda) [inline protocol/events.go]: Rename to SnapshotsBulk?

File: raftwal/commands/command.go [hunk 6]
```diff
@@ -27,15 +27,20 @@ import (
...
+	AddBulkEventCommandType
+	MetadataSetCommandType
+	MetadataDeleteCommandType
 )
 
 type AddEventCommand struct {
 	Event []byte
 }
 
+type AddBulkEventCommand struct {
```
Comment (aalda) [inline raftwal/commands/command.go]: Rename to AddEventsBulkCommand

File: raftwal/fsm.go [hunk 7]
```diff
@@ -115,7 +120,7 @@ func (s fsmState) shouldApply(f *fsmState) bool {
 		return false
 	}
 
-	if f.BalloonVersion > 0 && s.BalloonVersion != (f.BalloonVersion-1) {
+	if f.BalloonVersion > 0 && s.BalloonVersion >= f.BalloonVersion {
```
Comment (aalda) [inline raftwal/fsm.go]: Not sure about this. We should discuss it in person...

File: balloon/balloon.go (hunk 2 above)
Comment (iknite) [reply balloon/balloon.go]: We want to be homogeneus with the design, like `Events` struct (for example)

File: api/apihttp/apihttp.go [hunk 8]
```diff
@@ -97,24 +97,82 @@ func Add(balloon raftwal.RaftBalloonApi) http.HandlerFunc {
 			return
 		}
 
-		snapshot := &protocol.Snapshot{
-			response.HistoryDigest,
-			response.HyperDigest,
-			response.Version,
-			response.EventDigest,
-		}
+		snapshot := protocol.Snapshot(*(response))
```
Comment (jllucas) [inline api/apihttp/apihttp.go]: delete ()

File: balloon/balloon.go (hunk 2 above)
Comment (jllucas) [reply balloon/balloon.go]: type snapshotBulk []*Snapshots

File: raftwal/fsm_test.go [hunk 9]
```diff
@@ -14,32 +14,56 @@ import (
 	storage_utils "github.com/bbva/qed/testutils/storage"
 )
 
-func TestApply(t *testing.T) {
+func TestApplyAdd(t *testing.T) {
 
-	log.SetLogger("TestApply", log.SILENT)
+	log.SetLogger("TestApplyAdd", log.SILENT)
 
 	store, closeF := storage_utils.OpenRocksDBStore(t, "/var/tmp/balloon.test.db")
```
Comment (jllucas) [inline raftwal/fsm_test.go]: Change RocksDB for BPlus

File: raftwal/fsm_test.go [hunk 10]
```diff
@@ -153,12 +177,27 @@ func BenchmarkApplyAdd(b *testing.B) {
...
 }
 
-func newRaftLog(index, term uint64) *raft.Log {
+func newRaftAddLog(index, term uint64) *raft.Log {
 	event := []byte("All's right with the world")
 	data, _ := commands.Encode(commands.AddEventCommandType, &commands.AddEventCommand{Event: event})
 	return &raft.Log{Index: index, Term: term, Type: raft.LogCommand, Data: data}
 }
 
+func newRaftAddBulkLog(index, term uint64) *raft.Log {
```
Comment (jllucas) [inline raftwal/fsm_test.go]: Refactor newLog and newCommand

File: tests/e2e/add_bulk_verify_test.go [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+	"testing"
+
+	"github.com/bbva/qed/hashing"
+	"github.com/bbva/qed/protocol"
+	"github.com/bbva/qed/testutils/rand"
+	"github.com/bbva/qed/testutils/scope"
+	assert "github.com/stretchr/testify/require"
+)
+
+func TestAddBulkVerify(t *testing.T) {
```
Comment (jllucas) [inline tests/e2e/add_bulk_verify_test.go]: Rename

File: raftwal/raft.go [hunk 12]
```diff
@@ -375,19 +376,36 @@ func (b *RaftBalloon) Add(event []byte) (*balloon.Snapshot, error) {
...
+	}
+	b.metrics.Adds.Add(float64(len(bulk)))
+
+	snapshotBulk := resp.(*fsmAddBulkResponse).snapshotBulk
+
+	//Send snapshot to the snapshot channel
+	// TODO move this to an upper layer (shard manager?)
+	for _, s := range snapshotBulk.Snapshots {
+		p := protocol.Snapshot(s)
+		b.snapshotsCh <- &p
```
Comment (jllucas) [inline raftwal/raft.go]: Allow channel drainer in tests

File: raftwal/fsm.go (hunk 7 above)
Comment (aalda) [reply raftwal/fsm.go]: I think it's correct

File: raftwal/fsm.go [hunk 13]
```diff
@@ -141,6 +146,18 @@ func (fsm *BalloonFSM) Apply(l *raft.Log) interface{} {
 		}
 		return &fsmAddResponse{error: fmt.Errorf("state already applied!: %+v -> %+v", fsm.state, newState)}
 
+	case commands.AddBulkEventCommandType:
+		var cmd commands.AddBulkEventCommand
+		if err := commands.Decode(buf[1:], &cmd); err != nil {
+			return &fsmAddBulkResponse{error: err}
+		}
+		// INFO: after applying a bulk there will be a jump in term version due to balloon version mapping.
+		newState := &fsmState{l.Index, l.Term, fsm.balloon.Version()}
```
Comment (aalda) [inline raftwal/fsm.go]: The new fsmState should include the version that the balloon should have after applying the bulk, so instead `fsm.balloon.Version()` it should be `fsm.balloon.Version() + len(cmd.Events)-1`

**Pull request merged**